src/parser.ts

```diff
@@ -305,17 +306,63 @@ function parseSuperTypes(
   options: Options,
   processed: Processed,
   usedNames: UsedNames
-): TNamedInterface[] {
+): TSuperType[] {
   // Type assertion needed because of dereferencing step
   // TODO: Type it upstream
   const superTypes = schema.extends as SchemaSchema | SchemaSchema[] | undefined
   if (!superTypes) {
     return []
   }
   if (Array.isArray(superTypes)) {
-    return superTypes.map(_ => newNamedInterface(_, options, _, processed, usedNames))
+    return superTypes.map(_ => getOrCreateSuperType(_, options, _, processed, usedNames))
+  }
+  return [getOrCreateSuperType(superTypes, options, superTypes, processed, usedNames)]
+}
+
+function isSuperType(type: any): type is TSuperType {
+  return type.type === 'CUSTOM_TYPE' || type.type === 'INTERFACE'
+}
+
+function getOrCreateSuperType(
+  schema: SchemaSchema,
+  options: Options,
+  rootSchema: JSONSchema,
+  processed: Processed,
+  usedNames: UsedNames
+): TSuperType {
+  if (processed.has(schema)) {
+    const hit = processed.get(schema)
+    // Protect against programming mistakes
+    if (isSuperType(hit)) {
+      return hit
+    } else {
+      throw Error(format('type is not a supertype', hit))
+    }
+  }
+
+  if (schema.tsType) {
+    const extendedType = newExtendedType(schema as CustomTypeJSONSchema, usedNames)
+    processed.set(schema, extendedType)
+    return extendedType
+  }
+  const namedInterface = newNamedInterface(schema, options, rootSchema, processed, usedNames)
+  processed.set(schema, namedInterface)
+  return namedInterface
+}
+
+function newExtendedType(schema: CustomTypeJSONSchema, usedNames: UsedNames): TSuperType {
```

Comment:
Do you need this function, or can you call `parse` directly?

ie. in `newSupertype`:

```ts
const supertype = isCustomType(schema)
  ? parse(schema, options, keyName, processed, usedNames)
  : newNamedInterface(schema, options, rootSchema, processed, usedNames)
processed.set(schema, supertype)
return supertype
```

Reply:
I have no idea